packages/compartment-mapper/src/link.js

```diff
@@ -328,37 +210,110 @@ const makeModuleMapHook = (
+const isSyncOptions = value =>
+  !value ||
+  (typeof value === 'object' &&
+    !('moduleTransforms' in value) &&
+    'makeImportNowHook' in value &&
+    typeof value.makeImportNowHook === 'function');
+
 /**
  * Assemble a DAG of compartments as declared in a compartment map starting at
  * the named compartment and building all compartments that it depends upon,
  * recursively threading the modules exported by one compartment into the
  * compartment that imports them.
- * Returns the root of the compartment DAG.
- * Does not load or execute any modules.
- * Uses makeImportHook with the given "location" string of each compartment in
- * the DAG.
- * Passes the given globals and external modules into the root compartment
- * only.
  *
+ * - Returns the root of the compartment DAG.
+ * - Does not load or execute any modules.
+ * - Uses `makeImportHook` with the given "location" string of each compartment
+ *   in the DAG.
+ * - Passes the given globals and external modules into the root compartment
+ *   only.
+ *
+ * Differences between the two overloads:
+ *
+ * - _Must not_ accept `makeImportNowHook`.
+ * - _May_ accept `syncModuleTransforms`, but any keypair in
+ *   `moduleTransforms` (if provided) will overwrite it.
+ *
+ * @overload
  * @param {CompartmentMapDescriptor} compartmentMap
  * @param {LinkOptions} options
+ * @returns {LinkResult} the root compartment of the compartment DAG
+ */
+
+/**
+ * Assemble a DAG of compartments as declared in a compartment map starting at
+ * the named compartment and building all compartments that it depends upon,
+ * recursively threading the modules exported by one compartment into the
+ * compartment that imports them.
+ *
+ * - Returns the root of the compartment DAG.
+ * - Does not load or execute any modules.
+ * - Uses `makeImportHook` with the given "location" string of each compartment
+ *   in the DAG.
+ * - Passes the given globals and external modules into the root compartment
+ *   only.
+ *
+ * Differences between the two overloads:
+ *
+ * - _May_ accept `makeImportNowHook`.
+ * - _Must not_ accept `moduleTransforms`
+ * - _May_ accept `syncModuleTransforms`
+ *
+ * @overload
+ * @param {CompartmentMapDescriptor} compartmentMap
+ * @param {SyncLinkOptions} options
+ * @returns {LinkResult} the root compartment of the compartment DAG
+ */
+
+/**
+ * @param {CompartmentMapDescriptor} compartmentMap
+ * @param {LinkOptions|SyncLinkOptions} options
+ * @returns {LinkResult}
  */
 export const link = (
   { entry, compartments: compartmentDescriptors },
-  {
+  options,
+) => {
+  const {
     resolve = resolveFallback,
     makeImportHook,
     parserForLanguage: parserForLanguageOption = {},
     languageForExtension: languageForExtensionOption = {},
     globals = {},
     transforms = [],
-    moduleTransforms = {},
     __shimTransforms__ = [],
     archiveOnly = false,
     Compartment = defaultCompartment,
-  },
-) => {
+  } = options;
+
+  const isSync = isSyncOptions(options);
```

Comment:
Ideally `isSync` should not be needed in link. We could merge as-is but I want to refactor it before I forget.

Reply:
yes, let me take another look at that

Reply:
the logic has to go somewhere. I got rid of the variable, though.